common/proxy_connection_handler.rb

```diff
@@ -0,0 +1,61 @@
+# Connection Handler class for sending data to a Wavefront proxy.
+#
+# @author Yogesh Prasad Kurmi (ykurmi@vmware.com)
+
+require 'socket'
+
+require_relative 'atomic_integer'
+
+# Connection Handler.
+# For sending data to a Wavefront proxy listening on a given port.
+module Wavefront
+  class ProxyConnectionHandler
+    # Construct ProxyConnectionHandler.
+    # @param address [String] Proxy Address
+    # @param port [Integer] Proxy Port
+    def initialize(address, port)
+      @failures = AtomicInteger.new
+      @address = address
+      @port = port
+      @reconnecting_socket = nil
+    end
+
+    # Open a socket connection to the given address:port
+    def connect
```

Comment:
Have @lock = Mutex.new in the ProxyConnectionHandler
and then do this
def connect
   @lock.synchronize do
      @reconnecting_socket = TCPSocket.open(@address, @port)
   end